File: solutions/python/graphs/possible-bipartition.py
# LeetCode 886 — Possible Bipartition (Medium)
# Category: Graphs · Approach: BFS coloring
# Time: O(n + E) | Space: O(n + E)
# Source: https://leetcode.com/problems/possible-bipartition/

from collections import deque

def possible_bipartition(n: int, dislikes: list[list[int]]) -> bool:
    adj = [[] for _ in range(n + 1)]
    for a, b in dislikes:
        adj[a].append(b)
        adj[b].append(a)
    color = [0] * (n + 1)
    for s in range(1, n + 1):
        if color[s]:
            continue
        color[s] = 1
        queue = deque([s])
        while queue:
            u = queue.popleft()
            for v in adj[u]:
                if color[v] == 0:
                    color[v] = -color[u]
                    queue.append(v)
                elif color[v] == color[u]:
                    return False
    return True

File: solutions/typescript/graphs/possible-bipartition.ts
// LeetCode 886 — Possible Bipartition (Medium)
// Category: Graphs · Approach: BFS coloring
// Time: O(n + E) | Space: O(n + E)
// Source: https://leetcode.com/problems/possible-bipartition/

function possibleBipartition(n: number, dislikes: number[][]): boolean {
  const adj: number[][] = Array.from({ length: n + 1 }, () => []);
  for (const [a, b] of dislikes) {
    adj[a].push(b);
    adj[b].push(a);
  }
  const color = new Array(n + 1).fill(0);
  for (let s = 1; s <= n; s++) {
    if (color[s] !== 0) continue;
    color[s] = 1;
    const queue: number[] = [s];
    while (queue.length) {
      const u = queue.shift()!;
      for (const v of adj[u]) {
        if (color[v] === 0) {
          color[v] = -color[u];
          queue.push(v);
        } else if (color[v] === color[u]) {
          return false;
        }
      }
    }
  }
  return true;
}
